### ISSUE TITLE
CRITICAL BUG: Latest 1.10.2 release breaks the Auth0 package by not sending OAuth credentials

### ISSUE THREAD

**Issue by kopertop** [ngonzalvez/rest-facade#42]
The latest release completely breaks the Auth0 package, and no longer sends along any OAuth credentials.

Works fine with 1.10.1 but not with 1.10.2.

Still digging but reporting the issue here in case someone else has similar problems.

**Comment by kopertop** [comment]
Issue also cross posted on Auth0: https://github.com/auth0/node-auth0/issues/351

**Comment by ngonzalvez** [comment]
@kopertop thanks for letting me know. I'll take a look.

**Comment by ngonzalvez** [comment]
@kopertop I can't find what the issue, since I'm not being able to reproduce the bug. What is it that's not being sent? Is it data in the body of the request, in the headers, cookies, query-string? Could you be more specific?

**Comment by uditha-atukorala** [comment]
Not sure why the decision was made to do a patch release after some major upgrades to dependencies and I haven't had the change to look into where the actual error is.

At first glance I suspect it isn't setting the request headers as expected. Maybe because of the following change?
``` diff
-  this.options = deepmerge(defaultOptions, options || {}, true);
+  this.options = deepmerge(defaultOptions, options || {});
```

**Comment by ngonzalvez** [comment]
@uditha-atukorala before I did a patch release I checked the following things:

1. Before upgrading the dependencies I went to their docs and made sure that the breaking changes they introduced didn't affect **rest-facade**.
2. If they did change something that this library uses and the upgrade to the newer version of the dependency was requested, I made sure to update the corresponding code in **rest-facade**.
3. The automated tests are passing.

I've been double checking this, re-reading documentations, doing some manual testing, everything seems to be working. Even the change that you mention, that's the right thing to do. **deepmerge** no longer takes a boolean as a third argument, and it defaults the `clone` option to true.

Just in case, I just tested it, the options are merged as they should.

I checked the auth0 issue that was referenced above, it just says that there's something wrong with rest-facade. I'm trying my best, but without knowing how to reproduce the bug, I don't think I will be able to fix it soon.

**Comment by uditha-atukorala** [comment]
If you initialise a Auth0 management client (with an m2m client ID and client secret) and try to access the Auth0 API (e.g. get user info) Auth0 API returns 401 - Missing authentication. 

As I said before, I didn’t manage to track down where the exact error is but reverting to 1.10.1 solves the issue so assume it is an issue with 1.10.2 release. 

I’ll try to put some code together to reproduce the issue a bit later.

**Comment by ngonzalvez** [comment]
For the sake of posterity:

The `options` object you give to `rest-facade` is cloned to prevent the library from modifying the original reference. That's what the `true` in the following line was meant to do:

```javascript
this.options = deepmerge(defaultOptions, options || {}, true);
```
However, at some point, `deepmerge` deprecated that function signature. The third argument became an object, and the object cloning stopped working. `rest-facade` code wasn't updated to reflect this cange, so it kept the original reference and modified the original object. That was a bug.

This was fixed in the `v1.10.2` release. But it seems like the `node-auth0` relies on that bug and that breaks their authentication process. Since there's many people using the `node-auth0` library, I decided to revert the bug fix until their code is updated. Once that's done, I'll fix the object cloning feature and add the corresponding tests for it.
